src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -8,59 +8,47 @@ import com.openlattice.postgres.PostgresArrays
 import com.openlattice.postgres.PostgresTable.IDS
 import com.zaxxer.hikari.HikariDataSource
-import java.sql.Connection
 import java.time.OffsetDateTime
 import java.util.*
 
 
 class IndexingMetadataManager(private val hds: HikariDataSource, private val partitionManager: PartitionManager) {
 
     fun markAsIndexed(entityKeyIdsWithLastWrite: Map<UUID, Map<UUID, OffsetDateTime>>, linking: Boolean): Int {
-        var count = 0
-
         val entitySetPartitions = partitionManager.getEntitySetsPartitionsInfo(entityKeyIdsWithLastWrite.keys)
 
-        hds.connection.use { connection ->
-            entityKeyIdsWithLastWrite.forEach { (entitySetId, entities) ->
-                val partitions = entitySetPartitions.getValue(entitySetId).partitions.toList()
-
-                entities.entries
-                        .groupBy({ getPartition(it.key, partitions) }, { it.toPair() })
-                        .mapValues { it.value.toMap() }
-                        .forEach { (partition, entitiesWithLastWrite) ->
-
-                            entitiesWithLastWrite.entries
-                                    .groupBy { it.value }
-                                    .mapValues { it.value.map { it.key } }
-                                    .forEach { (lastWrite, entities) ->
-                                        count += markAsIndexed(
-                                                connection, entitySetId, entities, partition, lastWrite, linking
-                                        )
-                                    }
-                        }
+        return hds.connection.use { connection ->
+            val updateSql = if (linking) updateLastLinkingIndexSql else updateLastIndexSql
+
+            connection.prepareStatement(updateSql).use { stmt ->
+
+                entityKeyIdsWithLastWrite.forEach { (entitySetId, entities) ->
+
+                    val partitions = entitySetPartitions.getValue(entitySetId).partitions.toList()
+                    entities.entries
+                            .groupBy({ getPartition(it.key, partitions) }, { it.toPair() })
+                            .mapValues { it.value.toMap() }
+                            .forEach { (partition, entitiesWithLastWrite) ->
+
+                                entitiesWithLastWrite.entries
+                                        .groupBy { it.value }
+                                        .mapValues { it.value.map { it.key } }
+                                        .forEach { (lastWrite, entities) ->
+
+                                            val idsArray = PostgresArrays.createUuidArray(connection, entities)
+                                            stmt.setObject(1, lastWrite)
+                                            stmt.setObject(2, entitySetId)
+                                            stmt.setArray(3, idsArray)
+                                            stmt.setInt(4, partition)
+                                            stmt.addBatch()
```

Comment:
I think we also want to be setting partitions_version as a constraint here

Reply:
👍